src/context/AuthContext.jsx

```diff
@@ -0,0 +1,160 @@
+  const [loginCred, setLoginCred] = useState(initialLoginCredState);
+  const [signupCred, setSignupCred] = useState(initialSignupCredState);
+  const { callApi } = useApi();
+  const signUp = async (e) => {
+    e.preventDefault();
+    dispatch({ type: ACTION_TYPE_LOADING });
+    try {
+      if (signupCred.password !== signupCred.confirmPassword) {
+        setToast({
+          show: true,
+          content: "Passwords do not match",
+          type: "info",
+        });
+        dispatch({ type: ACTION_TYPE_SUCCESS, payload: [] });
+
+        return;
+      }
+
+      const result = await callApi("post", "auth/signup", false, {
+        name: signupCred.name,
+        email: signupCred.email,
+        password: signupCred.password,
+      });
+
+      setToast({
+        show: true,
+        content: `Welcome, ${result.data.name}`,
+        type: "info",
+      });
+      setLoginCred(initialLoginCredState);
+      setSignupCred(initialSignupCredState);
+      dispatch({ type: ACTION_TYPE_SUCCESS, payload: result.data });
+
+      setIsLoggedIn(true);
+    } catch (err) {
+      setToast({
+        show: true,
+        content: formatError(err),
+        type: "error",
+      });
+
+      dispatch({ type: ACTION_TYPE_FAILURE, payload: err.message });
+    }
+  };
+  const logIn = async (e) => {
+    e.preventDefault();
+    dispatch({ type: ACTION_TYPE_LOADING });
+    try {
+      const result = await callApi("post", "auth/login", false, {
+        email: loginCred.email,
+        password: loginCred.password,
+      });
+
+      setToast({
+        show: true,
+        content: `Welcome, ${result.data.name}`,
+        type: "info",
+      });
+      setLoginCred(initialLoginCredState);
+      setSignupCred(initialSignupCredState);
+      dispatch({ type: ACTION_TYPE_SUCCESS, payload: result.data });
+
+      setIsLoggedIn(true);
+    } catch (err) {
+      setToast({
+        show: true,
+        content: formatError(err),
+        type: "error",
+      });
+
+      dispatch({ type: ACTION_TYPE_FAILURE, payload: err.message });
+    }
+  };
+  const logOut = () => {
+    setToast({
+      show: true,
+      content: `Goodbye, ${state.data.name}`,
+      type: "warning",
+    });
+    localStorage.removeItem("token");
+    localStorage.removeItem("user");
+
+    setIsLoggedIn(false);
+    dispatch({ type: ACTION_TYPE_SUCCESS, payload: [] });
+  };
+  useEffect(() => {
+    if (isLoggedIn) {
+      setToken(localStorage.getItem("token"));
+      localStorage.setItem("user", JSON.stringify(state.data));
+      localStorage.setItem("token", state.data.token);
+      navigate("/", { replace: true });
+    }
+  }, [isLoggedIn]);
+  useEffect(() => {
+    if (localStorage.getItem("token")) {
+      setToken(localStorage.getItem("token"));
+      setIsLoggedIn(true);
+      dispatch({
+        type: ACTION_TYPE_SUCCESS,
+        payload: JSON.parse(localStorage.getItem("user")),
+      });
+
+      navigate("/", { replace: true });
```

Comment:
Why using **navigate in both the UseEffect**? Feels like some repeatative code. The former one will anyhow trigger when isLoggedIn changes